internal/state/execution.go

```diff
@@ -416,43 +414,29 @@ func buildLastCommitInfo(block *types.Block, store Store, initialHeight int64) a
 	}
 }
 
+//TODO reword
 // extendedCommitInfo expects a CommitInfo struct along with all of the
 // original votes relating to that commit, including their vote extensions. The
 // order of votes does not matter.
-func extendedCommitInfo(c abci.CommitInfo, votes []*types.Vote) abci.ExtendedCommitInfo {
-	if len(c.Votes) != len(votes) {
-		panic(fmt.Sprintf("extendedCommitInfo: number of votes from commit differ from the number of votes supplied (%d != %d)", len(c.Votes), len(votes)))
-	}
-	votesByVal := make(map[string]*types.Vote)
-	for _, vote := range votes {
-		if vote != nil {
-			valAddr := vote.ValidatorAddress.String()
-			if _, ok := votesByVal[valAddr]; ok {
-				panic(fmt.Sprintf("extendedCommitInfo: found duplicate vote for validator with address %s", valAddr))
-			}
-			votesByVal[valAddr] = vote
-		}
-	}
-	vs := make([]abci.ExtendedVoteInfo, len(c.Votes))
-	for i := range vs {
+func extendedCommitInfo(extCommit types.ExtendedCommit) abci.ExtendedCommitInfo {
+	vs := make([]abci.ExtendedVoteInfo, len(extCommit.ExtendedSignatures))
+	for i, ecs := range extCommit.ExtendedSignatures {
 		var ext []byte
-		// votes[i] will be nil if c.Votes[i].SignedLastBlock is false
-		if c.Votes[i].SignedLastBlock {
-			valAddr := crypto.Address(c.Votes[i].Validator.Address).String()
-			vote, ok := votesByVal[valAddr]
-			if !ok || vote == nil {
-				panic(fmt.Sprintf("extendedCommitInfo: validator with address %s signed last block, but could not find vote for it", valAddr))
-			}
-			ext = vote.Extension
+		if ecs.ForBlock() {
+			ext = ecs.VoteExtension
 		}
 		vs[i] = abci.ExtendedVoteInfo{
-			Validator:       c.Votes[i].Validator,
-			SignedLastBlock: c.Votes[i].SignedLastBlock,
+			Validator: abci.Validator{
+				Address: ecs.ValidatorAddress,
+				//TODO: Important: why do we need power here?
+				Power: 0,
```

Comment:
I left it there to not forget to write an issue to adapt the spec and the proto bugs so that we dont pastt "Validator" but just the validator's address.
Then I forgot 😊 
Doing it now...

Reply:
Done. As far as this PR is concerned, we can now follow @creachadair's suggestion

Reply:
The SDK uses this field to calculate the amount to slash. https://github.com/cosmos/cosmos-sdk/blob/90f61870e69b7cc07ac8537949715638af62f386/x/slashing/abci.go#L23

https://github.com/cosmos/cosmos-sdk/blob/90f61870e69b7cc07ac8537949715638af62f386/x/slashing/keeper/infractions.go#L87

If we plan to implement `SameBlock` execution, we should keep the full set of information here for parity between `FinalizeBlock` and `PrepareProposal`'s data.

Reply:
The parameter we are talking about here is not part of those that enable same-block. Note that this parameter (both in name and type) is different in `PrepareProposal`, `ProcessProposal`, and `FinalizeBlock`.

I looked into the links you indicate and indeed it looks like the SDK is getting the data from the request itself, but this seems overkill to me, since it is up to the App to decide the power of each validator (so it should know it offhand)... that's why it populates field `validator_updates` in `ResponseFinalizeBlock` (resp. in `ResponsePrepareProposal` and `ResponseProcessProposal` in same-block).

Reply:
We discussed this online and decided to not touch the spec, and populate the power in the same way we already do it in `ProcessProposal` and `FinalizeBlock`.
In that sense, @williambanfield was right in that it would enable same-block in the future

Reply:
Does [this](https://github.com/tendermint/tendermint/pull/8433/files#diff-e9bf9cb7440284b9dba54a6d42776bcfa91d60e0a29c9de1d5b1c872ba0ab6d0R422) address passing the validator power through to `PrepareProposal`?